test/e2e/main_test.go

```diff
@@ -138,6 +139,13 @@ func TestMain(m *testing.M) {
 				return ctx, err
 			}
 		}
+
+		log.Debugf("Creating namespace %v for tests", E2eNamespace)
+		nsObj := v1.Namespace{}
+		nsObj.Name = E2eNamespace
+		if err = cfg.Client().Resources().Create(ctx, &nsObj); err != nil {
+			return ctx, err
+		}
```

Comment:
We need make sure the created namespace is ready to use, eg:
```
		log.Infof("Wait for the %s namespace be ready", E2eNamespace)
		if err := wait.For(conditions.New(client.Resources()).ResourceMatch(nsObj, func(object k8s.Object) bool {
			ns, ok := object.(*v1.Namespace)
			if !ok {
				log.Printf("Not a namespace object: %v", object)
				return false
			}
			return ns.Status.Phase == corev1.NamespaceActive
		}), wait.WithTimeout(WAIT_NAMESPACE_AVAILABLE_TIMEOUT)); err != nil {
			return ctx, err
		}
		log.Infof("Namespace '%s' is ready for e2e-test", E2eNamespace)
```
otherwise we will got error looks like:
```
...
=== RUN   TestLibvirtCreateSimplePod/SimplePeerPod_test
    assessment_runner.go:223: pods "simple-test" is forbidden: error looking up service account e2e-test-8fb/default: serviceaccount "default" not found
--- FAIL: TestLibvirtCreateSimplePod (1.49s)
    --- FAIL: TestLibvirtCreateSimplePod/SimplePeerPod_test (1.49s)
...
```

Reply:
And do we need delete the created namespace at L173?
eg:
```
		// Delete the namespace for all test cases
		if nsObj != nil {
			client, err := cfg.NewClient()
			if err != nil {
				return ctx, err
			}
			if err = client.Resources().Delete(ctx, nsObj); err != nil {
				return ctx, err
			}
			log.Infof("Deleting namespace %s...", nsObj.Name)
			if err = wait.For(conditions.New(
				client.Resources()).ResourceDeleted(nsObj),
				wait.WithInterval(5*time.Second),
				wait.WithTimeout(60*time.Second)); err != nil {
				return ctx, err
			}
			log.Infof("Namespace %s has been successfully deleted within 60s", nsObj.Name)
		}
```

Reply:
Good point, we're still in a confusing (at least to me) place with respect to clean up. Based on the code it looks like there is a gap where we do teardown, but don't delete the cluster, so I think namespace deletion needs to be done in those circumstances, but they aren't especially common. It's still worth doing, so I'll look at adding that. Thanks for the suggestions!

Reply:
Hey @liudalibj - something that occurs to me after reading your good suggestions about improving the namespace function - I think rather than adding the create and wait for namespace and delete namespace to `main_tests.go`, now it's grown we are better to move it elsewhere for easier readability. I know you've done some refactoring recently, so wondered if you had any good suggestions? I was thinking `common.go`, but that's more peer pods and less Kubernetes type functions. Maybe assessment_helpers.go makes sense (though the name might want improving in future)?

Reply:
I decided on `assessment_helpers.go` in the short-term, but this can be changed before merging if required.

Reply:
Check the k8s e2e test codes, it seems that they create/delete namespace for every test case:
- create
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L260
- delete
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L353

So maybe we can use same logical? 
- add the e2e namespace object to [TestCase struct](https://github.com/confidential-containers/cloud-api-adaptor/blob/main/test/e2e/assessment_runner.go#L58)
- create and waiting the namespace at the top of [WithSetup](https://github.com/confidential-containers/cloud-api-adaptor/blob/main/test/e2e/assessment_runner.go#L174)
- delete the namespace at the end of [Teardown](https://github.com/confidential-containers/cloud-api-adaptor/blob/main/test/e2e/assessment_runner.go#L540)

Reply:
Create namespace once for all test cases and create namespace for everytime, both one is fine to me.

Reply:
Yeah, I saw those examples, but I don't want to create a unique namespace for each test case, but rather want them all shared for each run, so that we can easily delete all pods/services etc if we get a failure happening and want to clean things up. This is based on the approach that kata-containers moved to a while ago: https://github.com/kata-containers/kata-containers/pull/7238

Reply:
I am happy to discuss if people think that unique namespace per test, rather than test run is a better approach though. Sorry if I made it sound black and white. I was just implementing the issue description, but I'm also aware that I ghost-wrote that, so it's good too much of my thinking in already!

Reply:
I did abort a previous run in the middle of execution, but it seems like the error that you saw might not be related to the namespace not being in Active state:
```
time="2024-01-09T10:30:07Z" level=info msg="Creating namespace coco-pp-e2e-test-6afd3aac"
time="2024-01-09T10:30:07Z" level=info msg="Wait for namespace 'coco-pp-e2e-test-6afd3aac' be ready"
time="2024-01-09T10:30:12Z" level=info msg="Namespace 'coco-pp-e2e-test-6afd3aac' is ready"
=== RUN   TestLibvirtCreateSimplePod
=== RUN   TestLibvirtCreateSimplePod/SimplePeerPod_test
    assessment_runner.go:223: pods "simple-test" is forbidden: error looking up service account coco-pp-e2e-test-6afd3aac/default: serviceaccount "default" not found
--- FAIL: TestLibvirtCreateSimplePod (1.81s)
    --- FAIL: TestLibvirtCreateSimplePod/SimplePeerPod_test (1.81s)
=== RUN   TestLibvirtCreateSimplePodWithNydusAnnotation
=== RUN   TestLibvirtCreateSimplePodWithNydusAnnotation/SimplePeerPod_test
    assessment_runner.go:223: pods "alpine" is forbidden: error looking up service account coco-pp-e2e-test-6afd3aac/default: serviceaccount "default" not found
...
```
When I checked the namespace logic wasn't broken it seems to be okay:
```
# kubectl get namespaces
NAME                             STATUS   AGE
autorules                        Active   20h
coco-pp-e2e-test-6afd3aac        Active   68s
confidential-containers-system   Active   20h
default                          Active   20h
ingress-nginx                    Active   20h
kube-flannel                     Active   20h
kube-node-lease                  Active   20h
kube-public                      Active   20h
kube-system                      Active   20h
```
so there might be something else going on related to us creating the service account in another namespace?

I didn't see it on my early test runs though, so I'll try and uninstall the CAA and re-try in...